auslib/blobs/base.py

```diff
@@ -55,6 +55,42 @@ def createBlob(data):
     return blob_map[schema_version](**data)
 
 
+def merge_dicts(ancestor, left, right):
+    result = {}
+    for key in ancestor.keys() + left.keys() + right.keys():
+        key_types = set([type(ancestor.get(key)), type(left.get(key)), type(right.get(key))])
+        key_types.discard(type(None))
+        if len(key_types) > 1 and not key_types.issubset([str, unicode]):
+            raise ValueError("Cannot merge blobs: type mismatch for '{}'".format(key.encode('ascii', 'replace')))
+
+        if isinstance(ancestor.get(key), dict) or isinstance(left.get(key), dict) or isinstance(right.get(key), dict):
+            result[key] = {}
+            result[key] = merge_dicts(ancestor.get(key, {}), left.get(key, {}), right.get(key, {}))
+        elif isinstance(ancestor.get(key), list) or isinstance(left.get(key), list) or isinstance(right.get(key), list):
+            result[key] = ancestor.get(key, [])[:]
+            for side in (left, right):
+                for i in side.get(key, []):
+                    if i not in result[key] or not isinstance(i, type(result[key][result[key].index(i)])):
+                        result[key].append(i)
+        else:
+            if key in ancestor:
+                if key in left and key in right and ancestor[key] != left[key] and ancestor[key] != right[key]:
+                    raise ValueError("Cannot merge blobs: left and right are both changing '{}'".format(key.encode('ascii', 'replace')))
+                if key in left and ancestor[key] != left.get(key):
+                    result[key] = left[key]
+                else:
+                    result[key] = right[key]
+            else:
+                if key in left and key in right and left[key] != right[key]:
```

Comment:
Maybe this validation and validation at line 77 can be abstracted in another function.

Reply:
Hmmm, I'm not seeing how. Line 77 deals with ancestor, left, and right while this one only deals with left and right.

Reply:
Do you think a function with ancestor parameter none by default is good? So check ancestor only if it is not null.

Reply:
I might be missing something here, but I'm not sure how it will help. The two lines you mentioned test different things (ancestor vs left and ancestor vs right on line 77, left vs right on line 84).